[DTensor] Fix grouped_mm strategy for invalid stride cases

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158298
* __->__ #158245

local_tensor input to grouped_mm has a stride requirement.

(see `_meta_grouped_mm_common` in meta_registrations.py or
`check_valid_strides_and_return_transposed` in native/cuda/Blas.cpp)

Don't allow sharding a tensor if its shape would result in an
incompatible local_tensor stride.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k

wconstab committed:
[DTensor] Fix grouped_mm strategy for invalid stride cases

local_tensor input to grouped_mm has a stride requirement.

(see `_meta_grouped_mm_common` in meta_registrations.py or
`check_valid_strides_and_return_transposed` in native/cuda/Blas.cpp)

Don't allow sharding a tensor if its shape would result in an
incompatible local_tensor stride.

[ghstack-poisoned]

XilunWu reviewed (commented):
LGTM but I have some suggestions and also confusions to clarify.

wconstab commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -269,30 +278,36 @@ def expand_to_full_mesh_op_strategy(
+        output_specs: tuple[Optional[DTensorSpec], ...]
+        if input_index > 1:
+            output_specs = tuple(spec_list[:input_index])
+        else:
+            if spec_list[0] is not None:
+                output_specs = spec_list[0]  # type: ignore[assignment]
+            else:
+                raise RuntimeError("output spec is None")
```
FYI i just moved this logic from below, so it is pre-existing logic you are complaining about.

That said, I agree its not well written. 
1) input_index >1 is being checked above, but ==1 is not being checked.  That means the else branch has to consider both ==0 and ==1.  (perhaps also <1, which would be bad).  It could be written more explicitly.

XilunWu replied:
I see, thanks for clarifying!

wconstab commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -1035,6 +1039,66 @@ def grouped_mm_strategy(op_schema: OpSchema) -> OpStrategy:
         )
 
+    def valid_grouped_mm_strides(
+        input_specs: list[DTensorSpec], output_specs: tuple[Optional[DTensorSpec], ...]
```
no can do.  This function has to have a generic signature that is not specific to grouped_mm.  That's becuase this function's signature is defined by the API of `expand_to_full_mesh_op_strategy` which is a generic util that can be used by any op.

wconstab committed:
Update on "[DTensor] Fix grouped_mm strategy for invalid stride cases"


local_tensor input to grouped_mm has a stride requirement.

(see `_meta_grouped_mm_common` in meta_registrations.py or
`check_valid_strides_and_return_transposed` in native/cuda/Blas.cpp)

Don't allow sharding a tensor if its shape would result in an
incompatible local_tensor stride.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -1035,6 +1039,66 @@ def grouped_mm_strategy(op_schema: OpSchema) -> OpStrategy:
+                meta.shape, mesh, placements
+            )
+            return local_shape, local_stride, meta.dtype
+
```
fixed

zpcore commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -1035,6 +1039,51 @@ def grouped_mm_strategy(op_schema: OpSchema) -> OpStrategy:
         )
 
+    def valid_grouped_mm_strides(
+        input_specs: list[DTensorSpec], output_specs: tuple[Optional[DTensorSpec], ...]
```
`output_specs` is not used.

zpcore replied:
I see, you are trying to match with `is_valid_strategy_cb` pattern. Though it is not used here.

zpcore reviewed (approved):
LGTM!

wconstab committed:
Update on "[DTensor] Fix grouped_mm strategy for invalid stride cases"


local_tensor input to grouped_mm has a stride requirement.

(see `_meta_grouped_mm_common` in meta_registrations.py or
`check_valid_strides_and_return_transposed` in native/cuda/Blas.cpp)

Don't allow sharding a tensor if its shape would result in an
incompatible local_tensor stride.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

XilunWu reviewed (approved):
Thanks for addressing my question, the PR looks good to me.

wconstab commented:
@pytorchbot merge
